library/pk_wrap.c

```diff
@@ -85,18 +84,12 @@ static int rsa_verify_wrap( void *ctx, mbedtls_md_type_t md_alg,
         return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
 #endif /* SIZE_MAX > UINT_MAX */
 
-    if( sig_len < rsa_len )
-        return( MBEDTLS_ERR_RSA_VERIFY_FAILED );
```

Comment:
This changes the error returned in an error condition, and therefore breaks the ABI in an error condition. We do that from time to time, but only with good reason. In this case - what are the consequences and side effects, and are they justified? 

Until I know that, I don't think we can accept this PR, therefore tagging as needing design approval.

Reply:
Thanks @sbutcher-arm for the review. This callback is called by ```mbedtls_pk_verify()```. Condition ```mbedtls_pk_verify() != 0``` is considered non-recoverable failure at all places. No check for original error codes ```MBEDTLS_ERR_RSA_VERIFY_FAILED``` and ```MBEDTLS_ERR_PK_SIG_LEN_MISMATCH``` is done anywhere. These error codes only appear in debugs when printed in string form via ```mbedtls_strerror```.
Hence no impact on functionality. To test impact on testing I ran PR jobs [Jenkins](https://jenkins-internal.mbed.com/job/mbedtls-pr/1104/) and [Travis](https://travis-ci.org/ARMmbed/mbedtls/builds/345089780?utm_source=github_status&utm_medium=notification) and the [release job](https://jenkins-internal.mbed.com/job/mbedtls-release/306/). They all passed.

I like this change because lines [88](https://github.com/ARMmbed/mbedtls/pull/1149/files#diff-0e47986b5bc2042422a67b92ede10ee1L88) and  [96](https://github.com/ARMmbed/mbedtls/pull/1149/files#diff-0e47986b5bc2042422a67b92ede10ee1L96) effectively mean that ```sig_len``` should match ```rsa_len```. Hence the two conditions can be joined in one ```sig_len != rsa_len``` and more descriptive error ```MBEDTLS_ERR_PK_SIG_LEN_MISMATCH``` can be returned. As done in this PR.

Reply:
Yes adding a regression test would be a good idea. But that can be considered an enhancement.

Reply:
Commit 2abed842258aa714eb043cb717a75c10d1036cb2 made a change with this ChangeLog entry:

> pk_verify() now returns a specific error code when the signature is valid but shorter than the supplied length.

and the documentation of `MBEDTLS_ERR_PK_SIG_LEN_MISMATCH` states

> The signature is valid but its length is less than expected.

Changing the error code would therefore be an incompatible API change on paper. I do wonder why there is a specific error code for a signature that's too short. This error isn't treated differently inside the library and I can't think of a use case for it. The surrounding commits don't seem related either. @mpg Do you happen to remember the reason for this error code?

Reply:
Yes, it was required by some users (picocoin IIRC) that apparently had a use case where they don't know the exact size of the signature, only know an upper bound (ie the size of the buffer holding it, but there might be extra bytes after the signature in the buffer and they don't care about it). That specific error code was added so we could preserve the default behaviour of checking for an exact length match, while supporting them (they treat `MBEDTLS_PK_SIG_LEN_MISMATCH` as indicating a valid signature). As the documentation for this error code says: "The signature is valid but".

Reply:
Thanks Manuel! I think we should add a warning comment here then, and also clarify the error description. To me, a valid signature followed by trailing garbage is not a valid signature. Proposal in #1543